src/v8/isolate_scope.rs

```diff
@@ -82,17 +141,54 @@ impl<'isolate> V8IsolateScope<'isolate> {
     pub(crate) fn new_dummy(isolate: &'isolate V8Isolate) -> V8IsolateScope<'isolate> {
         V8IsolateScope {
             isolate,
-            inner_handlers_scope: std::ptr::null_mut(),
-            inner_isolate_scope: std::ptr::null_mut(),
+            storage: V8IsolateScopeStorage::default(),
         }
     }
 
+    /// Returns a raw pointer to the isolate.
+    pub fn get_raw_isolate(&self) -> *mut v8_isolate {
+        self.isolate.get_raw()
+    }
+
+    /// Creates a new [V8IsolateScope] (or "restores it") with an
+    /// already existing storage for the scopes.
+    /// The storage provided must have been created precisely for the
+    /// isolate passed.
+    pub fn restore(isolate: &'isolate V8Isolate, storage: V8IsolateScopeStorage) -> Option<Self> {
+        if storage.is_same_isolate(isolate) {
+            Some(Self { isolate, storage })
+        } else {
+            None
+        }
+    }
+
+    /// Forgets about the isolate and handlers scopes, so that those
+    /// aren't deleted in destructor. Returns the storage object created
+    /// for this [V8IsolateScope], so that it is still possible to
+    /// restore later.
+    pub fn forget(&mut self) -> V8IsolateScopeStorage {
+        std::mem::take(&mut self.storage)
+    }
+
+    /// Returns `true` if the scopes are left uninitialised.
+    pub fn is_forgotten(&self) -> bool {
+        self.storage.is_uninitialised()
+    }
+
     /// Creating a new context for JS code invocation.
-    #[must_use]
     pub fn new_context(&self, globals: Option<&V8LocalObjectTemplate>) -> V8Context {
         V8Context::new(self.isolate, globals)
     }
 
+    /// Returns a [V8ContextScope] if it has already been entered and
+    /// created for this isolate and isolate scope.
+    pub fn get_current_context_scope<'isolate_scope>(
```

Comment:
Would like to understand the need for this API as well

Reply:
I would still like to understand why we need this API `get_current_context_scope`. I an afraid it expose some unsafety. For example, this object can not outlive the `isolate_scope` and `isolate` but it can outlive the current `ContextScope` which is currently in use and exists somewhere. This means we might be able to continue working on the Context even after it is exit.

Reply:
If its only internal then lets just make it `pub(crate)`

Reply:
> The V8ContextScope has a method that prevents it from being freed after an object of V8ContextScope goes out of scope.

Which method is it?

I am OK with it been `pub(crate)` I see you already change it